packages/vite/src/node/plugins/css.ts

```diff
@@ -885,38 +921,79 @@ export function cssPostPlugin(config: ResolvedConfig): Plugin {
         })
       }
 
-      function extractCss() {
+      async function extractCss() {
         let css = ''
         const collected = new Set<OutputChunk>()
+        const concatCssEndLines: Array<{ file: string; end: number }> = []
         const prelimaryNameToChunkMap = new Map(
           Object.values(bundle)
             .filter((chunk): chunk is OutputChunk => chunk.type === 'chunk')
             .map((chunk) => [chunk.preliminaryFileName, chunk]),
         )
 
-        function collect(fileName: string) {
+        function prependImports(imports: string) {
+          if (!imports) {
+            return ''
+          }
+          return imports + '\n'
+        }
+
+        function collect(fileName: string): string {
           const chunk = bundle[fileName]
-          if (!chunk || chunk.type !== 'chunk' || collected.has(chunk)) return
+          if (!chunk || chunk.type !== 'chunk' || collected.has(chunk)) {
+            return ''
+          }
           collected.add(chunk)
 
-          chunk.imports.forEach(collect)
-          css += chunkCSSMap.get(chunk.preliminaryFileName) ?? ''
+          const css = chunkCSSMap.get(chunk.preliminaryFileName) ?? ''
+          const importedCSS = chunk.imports.reduce((css, file) => {
+            const imported = collect(file)
+
+            if (imported.length > 0) {
+              css = prependImports(css) + imported
+              concatCssEndLines.push({ file, end: getLineCount(css) })
+            }
+            return css
+          }, '')
+          return prependImports(importedCSS) + css
         }
 
-        for (const chunkName of chunkCSSMap.keys())
-          collect(prelimaryNameToChunkMap.get(chunkName)?.fileName ?? '')
+        for (const chunkName of chunkCSSMap.keys()) {
+          const filename = prelimaryNameToChunkMap.get(chunkName)?.fileName
+          if (!filename) {
+            continue
+          }
+
+          const cssCode = collect(filename)
 
-        return css
+          if (cssCode) {
+            concatCssEndLines.push({
+              file: filename,
+              end: getLineCount(cssCode),
+            })
+
+            css += cssCode
+          }
+        }
```

Comment:
Rethinking it, maybe this can be simplified to
```ts
        let line = 0

        function collect(fileName: string) {
          const chunk = bundle[fileName]
          if (!chunk || chunk.type !== 'chunk' || collected.has(chunk)) return
          collected.add(chunk)

          chunk.imports.forEach(collect)

          const content = chunkCSSMap.get(chunk.preliminaryFileName) ?? ''
          if (css !== '') {
            css += '\n' + content
          } else {
            css = content
          }
          line += getLineCount(content)
          concatCssEndLines.push({ file, end: line })
        }

        for (const chunkName of chunkCSSMap.keys())
          collect(prelimaryNameToChunkMap.get(chunkName)?.fileName ?? '')

        return css
      }
```

Reply:
I tried to implement your refactor here https://stackblitz.com/edit/node-xc3mmb?file=index.js.